From f50414e021c1baf568be6d2cfea3bbf88a57f80b Mon Sep 17 00:00:00 2001
From: nift4 <nift4@protonmail.com>
Date: Sun, 28 Aug 2022 13:35:56 +0200
Subject: [PATCH 29/34] InputManager: add cursor change listener

VNC can draw the cursor client-size if it can get a Bitmap from the
cursor. This allows us to fake a faster cursor and save server resources
because the framebuffer doesnt change when the cursor moves.

Change-Id: Ie6a9091205bf3848e82902a4862e0d837cdab402
---
 .../hardware/input/ICursorCallback.aidl       |  24 ++++
 .../android/hardware/input/IInputManager.aidl |   4 +
 .../android/hardware/input/InputManager.java  |  18 +++
 .../server/input/InputManagerService.java     | 110 ++++++++++++++++++
 4 files changed, 156 insertions(+)
 create mode 100644 core/java/android/hardware/input/ICursorCallback.aidl

diff --git a/core/java/android/hardware/input/ICursorCallback.aidl b/core/java/android/hardware/input/ICursorCallback.aidl
new file mode 100644
index 000000000000..ac1d02f0cbc3
--- /dev/null
+++ b/core/java/android/hardware/input/ICursorCallback.aidl
@@ -0,0 +1,24 @@
+/*
+ * Copyright (C) 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package android.hardware.input;
+
+import android.view.PointerIcon;
+
+/** @hide */
+interface ICursorCallback {
+    oneway void onCursorChanged(int iconId, in PointerIcon icon);
+}
diff --git a/core/java/android/hardware/input/IInputManager.aidl b/core/java/android/hardware/input/IInputManager.aidl
index 19fb845d9384..e43dfa66c4ad 100644
--- a/core/java/android/hardware/input/IInputManager.aidl
+++ b/core/java/android/hardware/input/IInputManager.aidl
@@ -25,6 +25,7 @@ import android.hardware.input.TouchCalibration;
 import android.os.CombinedVibration;
 import android.hardware.input.IInputSensorEventListener;
 import android.hardware.input.InputSensorInfo;
+import android.hardware.input.ICursorCallback;
 import android.hardware.lights.Light;
 import android.hardware.lights.LightState;
 import android.os.IBinder;
@@ -144,4 +145,7 @@ interface IInputManager {
     void openLightSession(int deviceId, String opPkg, in IBinder token);
 
     void closeLightSession(int deviceId, in IBinder token);
+
+    void registerCursorCallback(ICursorCallback callbacks);
+    void unregisterCursorCallback(ICursorCallback callbacks);
 }
diff --git a/core/java/android/hardware/input/InputManager.java b/core/java/android/hardware/input/InputManager.java
index 17116e248646..c62f20ad02f0 100644
--- a/core/java/android/hardware/input/InputManager.java
+++ b/core/java/android/hardware/input/InputManager.java
@@ -1141,6 +1141,24 @@ public final class InputManager {
         }
     }
 
+    /** @hide */
+    public void registerCursorCallback(ICursorCallback callback) {
+        try {
+            mIm.registerCursorCallback(callback);
+        } catch (RemoteException ex) {
+            throw ex.rethrowFromSystemServer();
+        }
+    }
+
+    /** @hide */
+    public void unregisterCursorCallback(ICursorCallback callback) {
+        try {
+            mIm.unregisterCursorCallback(callback);
+        } catch (RemoteException ex) {
+            throw ex.rethrowFromSystemServer();
+        }
+    }
+
     /**
      * Request or release pointer capture.
      * <p>
diff --git a/services/core/java/com/android/server/input/InputManagerService.java b/services/core/java/com/android/server/input/InputManagerService.java
index 1a6f785639c1..98d1ad36b2f2 100644
--- a/services/core/java/com/android/server/input/InputManagerService.java
+++ b/services/core/java/com/android/server/input/InputManagerService.java
@@ -44,6 +44,7 @@ import android.graphics.Point;
 import android.graphics.Rect;
 import android.hardware.display.DisplayManager;
 import android.hardware.display.DisplayViewport;
+import android.hardware.input.ICursorCallback;
 import android.hardware.input.IInputDevicesChangedListener;
 import android.hardware.input.IInputManager;
 import android.hardware.input.IInputSensorEventListener;
@@ -264,6 +265,12 @@ public class InputManagerService extends IInputManager.Stub
     @GuardedBy("mLidSwitchLock")
     private List<LidSwitchCallback> mLidSwitchCallbacks = new ArrayList<>();
 
+    private final Object mCursorCbLock = new Object();
+    @GuardedBy("mCursorCbLock")
+    private final SparseArray<CursorCallbackRecord> mCursorCallbacks = new SparseArray<>();
+    private final List<CursorCallbackRecord> mTempCursorCallbacksToNotify =
+            new ArrayList<>();
+
     // State for the currently installed input filter.
     final Object mInputFilterLock = new Object();
     IInputFilter mInputFilter; // guarded by mInputFilterLock
@@ -494,6 +501,75 @@ public class InputManagerService extends IInputManager.Stub
         }
     }
 
+    @Override // Binder call
+    public void registerCursorCallback(ICursorCallback listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+
+        synchronized (mCursorCbLock) {
+            final int callingPid = Binder.getCallingPid();
+            if (mCursorCallbacks.get(callingPid) != null) {
+                throw new IllegalStateException("The calling process has already registered "
+                        + "a CursorCallback.");
+            }
+            CursorCallbackRecord record =
+                    new CursorCallbackRecord(callingPid, listener);
+            try {
+                IBinder binder = listener.asBinder();
+                binder.linkToDeath(record, 0);
+            } catch (RemoteException ex) {
+                throw new RuntimeException(ex);
+            }
+            mCursorCallbacks.put(callingPid, record);
+        }
+    }
+
+    @Override // Binder call
+    public void unregisterCursorCallback(ICursorCallback listener) {
+        if (DEBUG) {
+            Slog.d(TAG, "unregisterCursorCallback: listener=" + listener + " callingPid="
+                    + Binder.getCallingPid());
+        }
+
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+
+        synchronized (mCursorCbLock) {
+            int callingPid = Binder.getCallingPid();
+            if (mCursorCallbacks.get(callingPid) != null) {
+                CursorCallbackRecord record = mCursorCallbacks.get(callingPid);
+                if (record.getListener().asBinder() != listener.asBinder()) {
+                    throw new IllegalArgumentException("listener is not registered");
+                }
+                mCursorCallbacks.remove(callingPid);
+            }
+        }
+    }
+
+    private void onCursorCallbackDied(int pid) {
+        synchronized (mCursorCbLock) {
+            mCursorCallbacks.remove(pid);
+        }
+    }
+
+    // Must be called on handler
+    private void deliverCursorChanged(int iconId, PointerIcon icon) {
+        mTempCursorCallbacksToNotify.clear();
+        final int numListeners;
+        synchronized (mCursorCbLock) {
+            numListeners = mCursorCallbacks.size();
+            for (int i = 0; i < numListeners; i++) {
+                mTempCursorCallbacksToNotify.add(
+                        mCursorCallbacks.valueAt(i));
+            }
+        }
+        for (int i = 0; i < numListeners; i++) {
+            mTempCursorCallbacksToNotify.get(i).notifyCursorChanged(iconId, icon);
+        }
+    }
+
     public void start() {
         Slog.i(TAG, "Starting input manager");
         nativeStart(mPtr);
@@ -2350,6 +2426,7 @@ public class InputManagerService extends IInputManager.Stub
     // Binder call
     @Override
     public void setPointerIconType(int iconId) {
+        deliverCursorChanged(iconId, null);
         nativeSetPointerIconType(mPtr, iconId);
     }
 
@@ -2357,6 +2434,7 @@ public class InputManagerService extends IInputManager.Stub
     @Override
     public void setCustomPointerIcon(PointerIcon icon) {
         Objects.requireNonNull(icon);
+        deliverCursorChanged(PointerIcon.TYPE_CUSTOM, icon);
         nativeSetCustomPointerIcon(mPtr, icon);
     }
 
@@ -3468,6 +3546,38 @@ public class InputManagerService extends IInputManager.Stub
         }
     }
 
+    private final class CursorCallbackRecord implements DeathRecipient {
+        private final int mPid;
+        private final ICursorCallback mListener;
+
+        public CursorCallbackRecord(int pid, ICursorCallback listener) {
+            mPid = pid;
+            mListener = listener;
+        }
+
+        @Override
+        public void binderDied() {
+            if (DEBUG) {
+                Slog.d(TAG, "Cursor listener for pid " + mPid + " died.");
+            }
+            onCursorCallbackDied(mPid);
+        }
+        
+        public ICursorCallback getListener() {
+            return mListener;
+        }
+
+        public void notifyCursorChanged(int iconId, PointerIcon icon) {
+            try {
+                mListener.onCursorChanged(iconId, icon);
+            } catch (RemoteException ex) {
+                Slog.w(TAG, "Failed to notify process " + mPid +
+                        " that cursor changed, assuming it died.", ex);
+                binderDied();
+            }
+        }
+    }
+
     private void onSensorEventListenerDied(int pid) {
         synchronized (mSensorEventLock) {
             mSensorEventListeners.remove(pid);
-- 
2.43.0

